File: lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.java

/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode(int x) { val = x; }
 * }
 */

import java.util.Stack;

class Solution {
    public TreeNode lowestCommonAncestor(TreeNode root, TreeNode p, TreeNode q) {
        Stack<TreeNode> st = new Stack<TreeNode>();
        Stack<TreeNode> st2 = new Stack<TreeNode>();
        st.push(root);
        st2.push(root);
        while (st.peek() != null && st.peek().val != p.val) {
            if (st.peek().val == q.val)
                return st.peek();
            if (st.peek().val > p.val)
                st.push(st.peek().left);
            else
                st.push(st.peek().right);
        }
        //System.out.println('1');
        while (st2.peek() != null && st2.peek().val != q.val) {
            if (st2.peek().val == p.val)
                return st2.peek();
            if (st2.peek().val > q.val)
                st2.push(st2.peek().left);
            else
                st2.push(st2.peek().right);
        }
        //System.out.println('2');
        while (st.size() < st2.size())
            st2.pop();
        while (st2.size() < st.size())
            st.pop();
        while (!st.empty() && !st2.empty() && st.peek().val != st2.peek().val) {
            st.pop();
            st2.pop();
        }
        return (st.empty() ? st2.peek() : st.peek());
    }
}

File: lowest-common-ancestor-of-a-binary-search-tree/lowest-common-ancestor-of-a-binary-search-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode(int x) : val(x), left(NULL), right(NULL) {}
 * };
 */
#define tn TreeNode
class Solution {
public:
    void alpha(tn *rt, stack<tn *> &s, int x)

{

	s.push(rt);

	if (rt->val == x)

		return;

	if (rt->val < x)

		return alpha(rt->right, s, x);

	return alpha(rt->left, s, x);

}
    TreeNode* lowestCommonAncestor(TreeNode* rt, TreeNode* a, TreeNode* b) {
        if (rt == NULL or (rt->left == NULL and rt->right == NULL))

	return rt;

stack<tn *> s1, s2;

alpha(rt, s1, a->val);

alpha(rt, s2, b->val);

while (s1.size() != s2.size())

	(s1.size() < s2.size() ? s2 : s1).pop();

tn *res = rt;

while (!s1.empty())

{

	tn *x = s1.top(), *y = s2.top();

	s1.pop();

	s2.pop();

	if (x->val == y->val)

	{

		res = x;

		break;

	}

}

return res;
    }
};